Make data streams in APIs resolvable.

martijnvg opened the pull request:
This change makes data stream resolvable in APIs.

The `INCLUDE_DATA_STREAMS` indices option controls whether data streams can be resolved in an api. In this pr, the `INCLUDE_DATA_STREAMS` indices option is enabled in the following APIs: search, msearch, refresh, index (op_type create only) and bulk (index requests with op type create only). In a subsequent later change, we will determine which other APIs need to be able to resolve data streams and enable the `INCLUDE_DATA_STREAMS` indices option for these APIs.

Whether an api resolve all backing indices of a data stream or the latest index of a data stream (write index) depends on the `IndexNameExpressionResolver.Context.isResolveToWriteIndex()`.
If `isResolveToWriteIndex()` returns true then data streams resolve to the latest index (for example: index api) and otherwise a data stream resolves to all backing indices of a data stream (for example: search api).

Relates to #53100

martijnvg committed:
Make data streams in APIs resolvable.

This change makes data stream resolvable in all APIs.
For APIs that have `FORBID_ALIASES_TO_MULTIPLE_INDICES` set
in the index options, this means that a data stream is resolved
to the latest backing index. For all other APIs this means
that a data stream is resolved to all backing indices.

Additionally also a `IGNORE_DATA_STREAMS` indices option is added,
so that for specific APIs data stream resolvability can be disabled.

Relates to #53100

elasticmachine commented:
Pinging @elastic/es-core-features (:Core/Features/Data streams)

martijnvg committed:
iter

martijnvg committed:
Merge remote-tracking branch 'es/master' into make_data_streams_resolvable_in_apis

martijnvg committed:
fixed checkstyle errors.

martijnvg committed:
fixed tests

martijnvg committed:
fixed forbidden api violation

martijnvg committed:
iter

martijnvg committed:
fixed checkstyle violation

martijnvg committed:
iter

martijnvg committed:
added more tests

martijnvg committed:
Merge remote-tracking branch 'es/master' into make_data_streams_resolvable_in_apis

martijnvg committed:
fixed test

martijnvg committed:
reuse test method

martijnvg committed:
altered test and fixed checkstyle violation

martijnvg committed:
Change node command so that it can parse, DataStreamMetadata.

It is used in the Metadata class and otherwise
node commands can't load the cluster state.

martijnvg committed:
added test

martijnvg committed:
Merge remote-tracking branch 'es/master' into make_data_streams_resolvable_in_apis

martijnvg committed:
supply core custom metadata to namedXContentRegistry of es node commands.

martijnvg committed:
Merge remote-tracking branch 'es/master' into make_data_streams_resolvable_in_apis

martijnvg committed:
iter

danhermann commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:
```diff
@@ -260,6 +272,11 @@ private static IllegalArgumentException aliasesNotSupportedException(String expr
 
+    private static IllegalArgumentException dataStreamsNotSupportedException(String expression) {
+        return new IllegalArgumentException("The provided expression [" + expression + "] matches a " +
+            "data stream, specify the corresponding concrete indices instead.");
```
Should the error message be updated to suggest specifying either concrete indices or aliases since either are legally resolvable if only `IGNORE_DATA_STREAMS` is specified?

danhermann commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:
```diff
@@ -766,9 +785,20 @@ private static void validateAliasOrIndex(String expression) {
 
         private static boolean aliasOrIndexExists(IndicesOptions options, Metadata metadata, String expression) {
```
It looks like there's a lot of places in this class where `aliasOrIndex` could be changed to `indexAbstraction`. Maybe something for a follow-up PR.

martijnvg replied:
Yes, we should have a follow up pr that replaces `aliasOrIndex` with `indexAbstraction`.

danhermann reviewed (commented):
Really cool to see it work with ingestion and search. I left a few comments though most are for follow-ups.

martijnvg commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexAbstraction.java:
```diff
@@ -257,4 +257,42 @@ private boolean isNonEmpty(List<IndexMetadata> idxMetas) {
+                          List<IndexMetadata> dataStreamIndices, IndexMetadata writeIndex) {
+            this.dataStream = dataStream;
+            this.dataStreamIndices = dataStreamIndices;
+            this.writeIndex = writeIndex;
```
Yes, agreed. If that pr lands before this pr gets merged then I will adjust this pr and otherwise the pr that adds the generation should also adjust the code here.

henningandersen reviewed (commented):
I did a first shallow iteration. My main comment is that with this "quick-fix" we enable this everywhere, including allowing all index related APIs to accept a data-stream name. This was questionable for aliases (for instance deleting indices through aliases), but I think it is more so for data-streams (asking to delete all indices of a data-stream through a data-stream seems strange).

I think we should rather make this opt-in to the individual APIs such that most of them fail out with data-stream not supported.

martijnvg commented on server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:
```diff
@@ -1369,6 +1369,25 @@ public Metadata build() {
+                    IndexMetadata writeIndex = backingIndices.get(backingIndices.size() - 1);
+                    IndexAbstraction existing = aliasAndIndexLookup.put(dataStream.getName(),
+                        new IndexAbstraction.DataStream(dataStream, backingIndices, writeIndex));
+                    if (existing != null) {
```
The line below throws an exception, so tests do fail if existing is not null?

martijnvg commented:
> I think we should rather make this opt-in to the individual APIs such that most of them fail out with data-stream not supported.

Agreed, I will change the `IGNORE_DATA_STREAMS` indices option to become `INCLUDE_DATA_STREAMS` and enable that in a few places, so that the data stream integration tests pass.

martijnvg committed:
applied first feedback

martijnvg committed:
renamed variable

martijnvg committed:
Merge remote-tracking branch 'es/master' into make_data_streams_resolvable_in_apis

martijnvg committed:
removed unused import

martijnvg committed:
adjust delete data stream logic to play nice with added assertions

martijnvg committed:
Only attempt to build indicesLookup if there are indices, otherwise data streams can't point to a concrete index.
Also fixed a few tests.

martijnvg committed:
fixed test due to error message change

martijnvg committed:
if yaml test runs in a mixed cluster then `Alias` may be returned due
to a change in the error message, this pr now returns `alias` with
originates from the index abstraction type's display name.

martijnvg committed:
clean data streams in test

martijnvg committed:
Replaced IGNORE_DATA_STREAMS indices option with INCLUDE_DATA_STREAMS indices option.

Data streams are not resolved unless the INCLUDE_DATA_STREAMS indices option is specified.
Only search, msearch, index (with op type create) and refresh requests are configured with
the INCLUDE_DATA_STREAMS indices option. All other requests / apis don't resolve data streams.

martijnvg committed:
fixed tests

martijnvg committed:
Merge remote-tracking branch 'es/master' into make_data_streams_resolvable_in_apis

martijnvg committed:
don't serialize include_datastreams if other node doesn't support it

danhermann reviewed (approved):
The latest PR updates LGTM

henningandersen commented:
I think we should we add a few REST tests too that checks that doing index admin ops against a data-stream fails? Like we have also for aliases, for instance when deleting an index. I am not sure we need full coverage on that, but would be good to ensure a handful of most important cases are handled, like delete index, update settings, update mapping, GET index (should that fail, I think so?).

martijnvg committed:
Merge remote-tracking branch 'es/master' into make_data_streams_resolvable_in_apis

martijnvg committed:
add yaml tests that check that using a data stream in a number of apis is forbidden.

henningandersen reviewed (commented):
@martijnvg let me know your thoughts on moving the `includeDataStreams` option from `IndicesOptions` to `IndexNameExpressionResolver` and failing always when hitting data-stream (and not wanting to resolve them)?

martijnvg commented on server/src/main/java/org/elasticsearch/action/support/IndicesOptions.java:
```diff
@@ -103,7 +102,8 @@ public static XContentBuilder toXContent(EnumSet<WildcardStates> states, XConten
         FORBID_CLOSED_INDICES,
-        IGNORE_THROTTLED;
+        IGNORE_THROTTLED,
+        INCLUDE_DATA_STREAMS;
```
> It looks like only search, refresh (and index) really need to resolve data streams?

I think more apis will need to be able to resolve data streams. I just enabled the ability for the APIs you mentioned in order to make integration tests pass. For example I think that field capability, get field mapping and eql APIs will need to resolve data streams as well. 

Also I suspect that get mapping and settings apis will need to be able to resolve data streams too (if a data stream is used then all mappings/settings for all backing indices are returned.) When talking to the Fleet team, they would like to the put mapping/settings apis to understand data streams as well (in this case a data stream should resolve to the latest index, otherwise they will need to figure out what the current write index is and do the required update). This was just part of casual conversation, so no decisions made yet.

I'm pretty sure that more apis will need to support data streams, so 
I prefer if we keep the current approach that enables the use data streams via an `Indices.Option` enum instance. Also the way I look at data streams is that it is a way to group indices and from that perspective exposing the ability to use data streams in apis as an indices option makes sense. I don't think we should ever hide the fact that a data stream is backed by indices. 

Only part of indices option is user specified. The include data streams indices option is not a user specified option as apposed to example allow_no_indices option. Each api/action can opt-in whether data streams should be resolvable. I think adding a method to `IndexNameExpressionResolver` will complicate the code. Now most apis use the same method for resolving user specified indices expressions and for all apis that need to be able support data streams this needs to be changed. Also apis tend to reuse the same base class.

henningandersen replied:
We spoke about this on another channel and came to the conclusion to keep this as is. There is agreement that a split of the user level and the action-specific options would be desirable, but we postpone that work to a separate effort.

martijnvg commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:
```diff
@@ -212,15 +220,21 @@
+                    concreteIndices.add(writeIndex.getIndex());
+                }
             } else {
                 if (indexAbstraction.getIndices().size() > 1 && !options.allowAliasesToMultipleIndices()) {
```
Whether an operation or api is expected to expand a wildcard expression to a data stream that depends on whether that api is setup to resolve data streams. So in this case that would mean that ignoring data streams is expected behaviour. This is at least my expectation.

martijnvg replied:
Henning and I chatted about this and we both now agree that the request should fail if data streams are not resolvable both the expression matches with one or more data streams. It makes it clear that an api doesn't support data streams (just like if a user specified a concrete data stream name) and this makes it easier to add data stream support if an api should support data streams.

martijnvg committed:
Merge remote-tracking branch 'es/master' into make_data_streams_resolvable_in_apis

martijnvg committed:
If include data stream indices option is not set and a wildcard expression expands to a data stream then fail.
This means the api doesn't support data streams, but the wildcard expression is resolving to a data stream.
Prior to this commit data streams would be quietly ignored.

We think that failing is more clear, so that it is clear that a specific api call isn't executed over a data stream.

martijnvg committed:
renamed test method

henningandersen commented on server/src/main/java/org/elasticsearch/action/support/IndicesOptions.java:
```diff
@@ -227,8 +237,21 @@ public boolean ignoreThrottled() {
-        out.writeEnumSet(options);
+        if (out.getVersion().before(Version.V_8_0_0) && options.contains(Option.INCLUDE_DATA_STREAMS)) {
+            EnumSet<Option> copy = EnumSet.copyOf(options);
+            copy.remove(Option.INCLUDE_DATA_STREAMS);
```
I think we should put in a check into CreateDataStreamAction to not allow creating data streams before all nodes are on the target release (7.8)? To ensure that there are no surprises in a mixed cluster.

martijnvg replied:
👍 I will make this change in a followup change.

henningandersen reviewed (approved):
LGTM

martijnvg committed:
renamed constant
